src/main/java/com/google/sps/Interactions.java

```diff
@@ -93,4 +105,70 @@ public static int dotProduct(Map<String, Integer> v1, Map<String, Integer> v2) {
     }
     return result;
   }
+
+  /**
+   * Creates an Interaction entity in datastore recording an interaction between user and event.
+   *
+   * @param userId the user's id as identified in datastore
+   * @param eventId the event's id as identified in datastore
+   * @param score the new rating by the user for the event
+   * @param forceOverride if false, only overwrites if new score > old score
+   * @return change of user's rating on an item (saves highest score only)
+   */
+  public static int recordInteraction(
+      String userId, long eventId, int score, boolean forceOverride) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query q =
+        new Query("Interaction")
+            .setFilter(
+                CompositeFilterOperator.and(
+                    new FilterPredicate("user", FilterOperator.EQUAL, userId),
+                    new FilterPredicate("event", FilterOperator.EQUAL, eventId)));
+    PreparedQuery pq = datastore.prepare(q);
+    Entity interactionEntity = null;
+    int delta = score;
+    try {
+      interactionEntity = pq.asSingleEntity();
+      int prevScore = Integer.parseInt(interactionEntity.getProperty("rating").toString());
+      if (forceOverride || prevScore < score) {
+        interactionEntity.setProperty("rating", score);
+        delta = score - prevScore;
+      } else {
+        delta = 0;
+      }
+    } catch (TooManyResultsException e) {
+      // delete existing entries and re-create later
```

Comment:
Please log whenever an exception is caught.

Reply:
Got it!